agent/plugins.go

```diff
@@ -35,17 +37,44 @@ func (a *Agent) setupPluginConfig() map[string][]*config.Plugin {
 		cfg[plugins.PluginTypeTarget] = a.config.Targets
 	}
 
-	// Iterate the configurations to match any nomad plugins. If these plugins
-	// have an empty config map, update this with the agents top level Nomad
-	// config. This helps stop the need for repeating the same params over and
-	// over.
+	// Iterate the configs and perform the config setup on each. If the
+	// operator did not specify any config, it will be nil so make sure we
+	// initialise the map.
 	for _, cfgs := range cfg {
 		for _, c := range cfgs {
-			if (c.Name == plugins.InternalAPMNomad || c.Name == plugins.InternalTargetNomad) && len(c.Config) == 0 {
-				c.Config = nomadHelper.ConfigToMap(a.config.Nomad)
+			if c.Config == nil {
+				c.Config = make(map[string]string)
 			}
+			a.setupPluginConfig(c.Config)
 		}
 	}
 
 	return cfg
 }
+
+// setupPluginConfig takes the individual plugin configuration and merges in
+// namespaced Nomad configuration unless the user has disabled this
+// functionality.
+func (a *Agent) setupPluginConfig(cfg map[string]string) {
+
+	// Look for the config flag that users can supply to toggle inheriting the
+	// Nomad config from the agent. If we do not find it, opt-in by default.
+	val, ok := cfg[plugins.ConfigKeyNomadConfigInherit]
+	if !ok {
+		nomadHelper.MergeMapWithAgentConfig(cfg, a.config.Nomad)
+		return
+	}
+
+	// Attempt to convert the string. If the operator made an effort to
+	// configure the key but got the value wrong, log the error and do not
+	// perform the merge. The operator can fix the error and we do not make an
+	// assumption.
+	boolVal, err := strconv.ParseBool(val)
+	if err != nil {
+		a.logger.Error("failed to convert config value to bool", "error", err)
+		return
+	}
```

Comment:
In the future I think we should probably validate the config file and exit on error. It's easy to miss a log line.

It would also lead a slightly better UX since we could point to the specific block that has the error.